(Re-)implement image_single

CyberShadow opened the pull request:
This fixes all graphical artifacts in Daniel Remar's Violet Metal.

`multifunction_variant` is inappropriate here because

1. `image_single` does not need to hold a copy of a variable; it just needs to clear a second variable (`image_speed`) when set
2. The type of the field is not `variant`, but `gm_scalar`, and changing it to `variant` breaks things elsewhere.

JoshDreamland commented:
I believe this actually works as of C++11. To clarify, do expressions like `image_single != -1` and `image_single + 5` work correctly? What about `var frame = image_single;`?

JoshDreamland commented on ENIGMAsystem/SHELL/Universal_System/graphics_object.h:
```diff
@@ -50,6 +50,11 @@ namespace enigma
+      struct {
+        object_graphics* parent;
+        void operator= (gs_scalar v) { parent->image_index = v; parent->image_speed = 0; }
+        operator gs_scalar() const { return parent->image_index; }
```
```C++
return parent->image_speed ? -1 : parent->image_index;
```

CyberShadow replied:
Done.

CyberShadow commented:
> believe this actually works as of C++11.

In what way? I don't think there are any C++11-isms here.

> To be sure, do expressions like image_single != -1 and image_single + 5 work correctly?

Looks like it.

> What about var frame = image_single;?

That one didn't work. I added a fix, but not sure about its correctness.

JoshDreamland commented:
>> believe this actually works as of C++11.
>
>In what way? I don't think there are any C++11-isms here.

I was mistaken; I thought this was taking advantage of an RFC I read a while back, but in actuality, it just has problems with, eg, `+=`, and I wasn't remembering the actual caveats of only implementing a cast operator and an assignment operator. I think there are also problems with saying, eg, `image_single = obj_player.image_single` (you would get the default `operator=`, which would lead to invalid accesses).

----

As discussed, I think the set of fixes you would need to add to make this work would be a bottomless chasm eventually converging on exactly what `multifunction_variant` implements, but without the extra string member. If you want to branch `multifunction_variant` but use `get()` and `set()` instead of giving it its own storage, that's fine, but I think that's a lot of work for such a simple feature, so feel free to just extend the existing `multifunction_variant` to make this work. Or just build your change into Violet, and we'll add this feature down the road.

RobertBColton commented:
If you could please also `git pull upstream master` before your next commit on this as well. That will give you the new Travis YAML so it won't take 2 days waiting on Mac.

CyberShadow commented:
OK, I rebased and switched the implementation to multifunction_variant. But, as discussed previously, it now fails to compile:

```
In file included from SHELLmain.cpp:99:0:
/home/vladimir/.enigma/Preprocessor_Environment_Editable/IDE_EDIT_evparent.h: In member function ‘virtual variant enigma::event_parent::myevent_beginstep()’:
/home/vladimir/.enigma/Preprocessor_Environment_Editable/IDE_EDIT_evparent.h:43:94: error: operands to ?: have different types ‘double’ and ‘variant’
   {xprevious = x; yprevious = y; if (sprite_index != -1) image_index = fmod((image_speed < 0)?(sprite_get_number(sprite_index) + image_index - fmod(abs(image_speed),sprite_get_number(sprite_index))):(image_index + image_speed), sprite_get_number(sprite_index));}
                                                                             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/vladimir/.enigma/Preprocessor_Environment_Editable/IDE_EDIT_evparent.h:43:94: note:   and each type can be converted to the other
```

Please advise.

CyberShadow commented:
> Or just build your change into Violet, and we'll add this feature down the road.

I would like to avoid redistributing patched game files themselves, as they are implicitly copyrighted and thus non-redistributable. So, I would like to get everything working without any modifications to the game itself. (It also has a higher long-term benefit, as fixing compatibility issues in ENIGMA rather than the particular game allows other games to benefit from the improved compatibility.)

JoshDreamland commented:
Oh, I'm sorry. Let me clarify: I didn't mean for you to have to use `multifunction_variant` for each of those values, as assigning to `image_speed` and `image_index` can never affect `image_single`. If you just make `image_single` the `multifunction_variant`, and have it zero `image_speed` and update `image_index` on assign, that should suffice. It will differ in behavior from GM only if `image_index` is updated explicitly after setting `image_single`, which I doubt there's a single game that does (it would just be overridden in Game Maker).

JoshDreamland commented on ENIGMAsystem/SHELL/Universal_System/graphics_object.cpp:
```diff
@@ -71,6 +79,19 @@ namespace enigma
+  }
+
+  void image_singlev::function(variant) {
+    *image_index = rval.d;
```
```C++
if (rval.d == -1) {
  *image_speed = 1;
} else {
  *image_index = rval.d;
  *image_speed = 0;
}
```

JoshDreamland commented on ENIGMAsystem/SHELL/Universal_System/graphics_object.cpp:
```diff
@@ -71,6 +79,19 @@ namespace enigma
   depthv::depthv() : myiter(0) {}
   depthv::~depthv() {}
 
+  void image_indexv::function(variant) {
```
Just lose these two.

JoshDreamland commented on ENIGMAsystem/SHELL/Universal_System/graphics_object.h:
```diff
@@ -45,12 +45,28 @@ namespace enigma
     depthv();
     ~depthv();
   };
+  struct image_indexv: multifunction_variant {
```
Drop `image_indexv`/`image_speedv`; revert 67/68 (52/53).

JoshDreamland commented:
In all fairness, the solution you had is probably more correct, but I think it wouldn't be worth the performance hit. If you feel better about it, though, you can fix that compilation error you're getting by using a cast to double. ENIGMA should handle that cast problem in user code.

CyberShadow commented:
Can we go back to the idea of refactoring `multifunction_variant`? That option seems to have the fewest compromises.

JoshDreamland commented:
That idea is sort of fraught with danger... the nice thing about `multifunction_variant` is that it extends `variant`, which is what we create special overloads for in the engine (and the compiler handles specially). Because `multifunction_variant` *is* a `variant`, that conversion is favored over any of the other applicable arguments. But there's no way to tell the C++ compiler that this new class would be a double or an integer. We would need to create special overloads for any function that we wanted to be able to pass these to.

Bottom line is, doing this correctly is a long-term change.

One thing I think we can do is create a set of template specializations called `enigma::offered<T>` and `enigma::preferred<T>`. The `offered` class would overload `operator T` and extend `preferred<T>`; `preferred<T>` would offer constructors for as many variations of `T` as possible. Functions would accept `prefered<final_type>`. I will have to play with this idea to see if it actually works; the point is, it's not a change we can cleanly make right now.

Another thing we can do is just define getter and setter operations as part of ENIGMA's compiler. That change is too difficult given ENIGMA's current infrastructure, but as I said, it's definitely on the roadmap.

For now, I think our best bet is `multifunction_variant`, even if it comes with some costs. I believe the minor quirk I mentioned above is perfectly acceptable to avoid the rest of the system having knowledge of `image_single`.

JoshDreamland commented:
Actually... thinking more about it, as long as `var` and `variant` knew about this new type, and that type only ever overloaded *one* cast operator, this could work. It could not serve as the basis of `var` or `variant`, but it could work alongside them. We can chat more about what that might look like, tonight.

CyberShadow commented:
Looks like I didn't catch you. Will you be online this weekend, maybe?

I dropped `multifunction_variant` for `image_index`/`image_speed` and reverted them to `gs_scalar` for now, which does compile and work.

RobertBColton commented:
@CyberShadow he's online more on the weekend actually, since he won't have work.
